app/src/main/java/com/example/fittracker/mvp/presenter/MainScreenPresenter.java

```diff
@@ -25,4 +40,43 @@ public void onBackPressed() {
     public void onLogOutPressed() {
         view.onLogOutPressed();
     }
+
+    @Override
+    public void checkLocationPermission() {
+        if (!view.locationPermissionGranted()) {
+            view.requestPermissions();
+        } else {
+            setLocationObject();
+        }
+    }
+
+    @SuppressLint("MissingPermission")
+    @Override
+    public void setLocationObject() {
+        model.setLocationObject(locationManager.getLastKnownLocation(LocationManager.GPS_PROVIDER));
+    }
+
+    public void init() {
+        locationManager = view.getLocationManager();
+        checkLocationPermission();
+        model.setWeatherService(WeatherGenerator.createService(WeatherService.class));
```

Comment:
please separate the weather call in a separate method and put a loader so the user will know that something is happening. Then, when the info will be setted you hide the loader